### DISCUSSION TITLE
PushButtons in player-sdfat-i2s

### DISCUSSION THREAD

**Discussion by carlesfg** [pschatzmann/arduino-audio-tools#1060] [Q&A] [answered]
Good morning,
I tested the player-sdfat-i2s program on a Wemos D1 R32 and it works very well.
Then I added 4 PushButtons. You have the [ image made with Fritzing here](https://xn--gnuscultura-dbb.eu/wemosAudio/player-sdfat-i2s-PushButtons.png).

You have the schema of  [Fritzing to modify here](https://xn--gnuscultura-dbb.eu/wemosAudio/player-sdfat-i2s-PushButtons.fzz).

I would like to try this program with my high school students from Valencia(Spain), the program works well when I press PushButton 1 (Play). 
The sequence of PushButton 3 (Next), PushButton 4 (Resume) and PushButton 1 (Play) also works but it only works once.
The Micro SD card has 12 mp3 songs from "Es ist Juli" that I cannot post here to avoid copyright infringement.

Can you check my modification to find the error?

You can download the program [program](https://xn--gnuscultura-dbb.eu/wemosAudio/player-sdfat-i2s-PushButtons.zip) here.

I post also here:

/**
 * @file player-sdfat-i2s.ino
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-sdfat-i2s/README.md
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3

 * Beispielcode Version Maerz2023 getestet von prilchen.de
 *
 * Modificat a l'IES Gabriel Ciscar d'Oliva (Spain) al novembre de 2023
 * @file player-sdfat-i2s-PushButtons.ino
 */

#include "AudioTools.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecMP3Helix.h"

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSDFAT source(startFilePath, ext);
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);

boolean Valor_Polsador_PlayerCopy;
boolean Valor_Polsador_Stop;
boolean Valor_Polsador_Next;
boolean Valor_Polsador_Play;
int pinPolsador_PlayerCopy = 26;
int pinPolsador_Stop = 25;
int pinPolsador_Next = 17;
int pinPolsador_Play = 16;
int EstatRetingut=0; // Guardem l'estat del reproductor // Save state of music player

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
  //AudioLogger::instance().begin(Serial, AudioLogger::Info); // Nur zum Testen diese Zeile Aktivieren, um Details im Serial Monitor zu sehen
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);  // Im Betrieb diese Zeile aktivieren

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);
  i2s.begin(cfg);

  // setup player
  player.setMetadataCallback(printMetaData);
  player.begin();
  pinMode(pinPolsador_PlayerCopy, INPUT);
  pinMode(pinPolsador_Stop, INPUT);
  pinMode(pinPolsador_Next, INPUT);
  pinMode(pinPolsador_Play, INPUT);
}

void loop() {
    Valor_Polsador_PlayerCopy = digitalRead(pinPolsador_PlayerCopy);
    Valor_Polsador_Stop = digitalRead(pinPolsador_Stop);
    Valor_Polsador_Next = digitalRead(pinPolsador_Next);
    Valor_Polsador_Play = digitalRead(pinPolsador_Play);
      

    
    if (Valor_Polsador_PlayerCopy == 1) {
      Serial.println("Engeguem la música // start music");
      EstatRetingut = 1;
    }//final de if1 // end of if1

    if (Valor_Polsador_Stop == 1) {
      Serial.println("Parem la música // Stop music");
      EstatRetingut = 2;
    }//final de if2 // end of if2

    if (Valor_Polsador_Next == 1) {
      Serial.println("Avança una cançó // move to the next file");
      EstatRetingut = 3;
    }//final de if3 // end of if3

    if (Valor_Polsador_Play == 1) {
      Serial.println("Reengeguem la música// Resuming music");
      EstatRetingut = 4;
    }//final de if4 // end of if4

switch (EstatRetingut) {
  case 0:
    // No farem res // Do nothing
    break;
  case 1:
    player.copy();// engeguem la música // start music
    break;
  case 2:
    player.stop();// Parem la música // Stop music
    break;
  case 3:
    player.next();// Avancem al següent // move to the next file
    break;
  case 4:
    player.play();// Reengeguem la música // Resuming music
    break; 
  }// final_switch_case // end__switch_case
}// final de programa // end of program

**Comment by pschatzmann** [comment, marked as answer]
I would not introduce separate variables for state management but query or let  the player handle it automatically: e.g. after calling stop(), you can still call copy: it's just not doing anything...
 
Whenever you poll pins in a loop and want to act on it you should debounce the result, otherwise you risk that the action is executed multiple times. 

You could use the [AudioActions](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_actions.html) class that I have added for handling the AudioKit buttons, but it is working with any pin as well. 

Here is an [example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sd-audiokit/player-sd-audiokit.ino) how to use it...

**Reply by carlesfg** [reply]
Ok, I'll try it.
Vielen Dank für Ihre Hilfe.
